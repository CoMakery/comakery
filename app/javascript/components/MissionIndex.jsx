import React from 'react'
import PropTypes from 'prop-types'
import {fetch as fetchPolyfill} from 'whatwg-fetch'
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd'
import Layout from './layouts/Layout'
import SidebarItem from './styleguide/SidebarItem'
import SidebarItemBold from './styleguide/SidebarItemBold'
import InputField from './styleguide/InputField'
import Icon from './styleguide/Icon'

export default class MissionIndex extends React.Component {
  constructor(props) {
    super(props)
    this.handleListClick = this.handleListClick.bind(this)
    this.updateMissionStatus = this.updateMissionStatus.bind(this)
    this.updateProjectStatus = this.updateProjectStatus.bind(this)
    this.onDragEnd = this.onDragEnd.bind(this)
    this.state = {
      selectedMissionId: null,
      missions         : props.missions
    }
  }

  onDragEnd(result) {
    // dropped outside the list
    if (!result.destination) {
      return
    }
    // same position
    if (result.source.index === result.destination.index) {
      return
    }
    // slice array
    const direction = result.source.index < result.destination.index ? -1 : 1
    const [startIndex, endIndex] = [Math.min(result.source.index, result.destination.index), Math.max(result.source.index, result.destination.index)]
    const subarray = this.state.missions.slice(startIndex, endIndex + 1)
    // need to update db

    fetchPolyfill(`/missions/rearrange`, {
      credentials: 'same-origin',
      headers    : {
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body  : JSON.stringify({
        mission_ids         : subarray.map(mission => mission.id),
        display_orders      : subarray.map(mission => mission.displayOrder),
        direction           : direction,
        'authenticity_token': this.props.csrfToken
      })
    }).then(response => {
      if (response.status === 200) {
        // need to update frontend
        response.json().then(data => {
          this.setState({
            missions: data.missions
          })
        })
      } else {
        response.json().then(data => {
          this.setState({
            errors      : data.errors,
            errorMessage: data.message
          })
        })
      }
    })
  }

  handleListClick(missionId) {
    this.setState({
      selectedMissionId: missionId
    })
  }

  updateMissionStatus(index) {
    const mission = this.state.missions[index]
    const status = mission.status === 'active' ? 'passive' : 'active'

    fetchPolyfill(`/missions/${mission.id}`, {
      credentials: 'same-origin',
      headers    : {
        'Content-Type': 'application/json'
      },
      method: 'PATCH',
      body  : JSON.stringify({
        mission             : {status: status},
        'authenticity_token': this.props.csrfToken
      })
    }).then(response => {
      if (response.status === 200) {
        let newMissions = Array.from(this.state.missions)
        newMissions[index].status = status
        this.setState({
          missions: newMissions
        })
      } else {
        response.json().then(data => {
          this.setState({
            errors      : data.errors,
            errorMessage: data.message
          })
        })
      }
    })
  }

  updateProjectStatus(index) {
    const { missions, selectedMissionId } = this.state
    if (selectedMissionId === null) {
      return // if no mission is selected
    }
    const selectedMissionIndex = missions.findIndex(mission => mission.id === selectedMissionId) // find index of the selected mission
    const project = missions[selectedMissionIndex].projects[index]
    const status = project.status === 'active' ? 'passive' : 'active'

    fetchPolyfill('/projects/update_status', {
      credentials: 'same-origin',
      headers    : {
        'Content-Type': 'application/json'
      },
      method: 'PATCH',
      body  : JSON.stringify({
        'project_id'        : project.id,
        'status'            : status,
        'authenticity_token': this.props.csrfToken
      })
    }).then(response => {
      if (response.status === 200) {
        let newMissions = Array.from(missions)
        newMissions[selectedMissionIndex].projects[index].status = status
        this.setState({
          missions: newMissions
        })
      } else {
        response.json().then(data => {
          this.setState({
            errors      : data.errors,
            errorMessage: data.message
          })
        })
      }
    })
  }

  render() {
    const { selectedMissionId, missions } = this.state
    const selectedMission = selectedMissionId !== null
      ? missions.find(mission => mission.id === selectedMissionId) : null

    return (
      <React.Fragment>
        <Layout
          className="mission-index"
          category="Missions > "
          title="Configure missionâ€™s landing page"
          sidebar={
            <React.Fragment>
              <div className="mission-index--sidebar">
                <SidebarItemBold
                  className="mission-index--sidebar--item__bold"
                  iconLeftName="MISSION/WHITE.svg"
                  iconRightName="PLUS.svg"
                  text="Create a Mission"
                  onClick={(_) => window.location = '/missions/new'}
                />

                { missions.length > 0 &&
                  <React.Fragment>
                    <hr />

                    <div className="mission-index--sidebar--info">
                      Please select or rearrange missions you want to display on landing page:
                    </div>

                    <DragDropContext onDragEnd={this.onDragEnd}>
                      <Droppable droppableId="droppable">
                        {(provided, _) => (
                          <div
                            ref={provided.innerRef}
                          >
                            {missions.map((mission, index) => (
                              <Draggable key={mission.id} draggableId={mission.id} index={index}>
                                {(provided, _) => (
                                  <div
                                    ref={provided.innerRef}
                                    {...provided.draggableProps}
                                    {...provided.dragHandleProps}
                                  >
                                    <SidebarItem
                                      className="mission-index--sidebar--item"
                                      key={mission.id}
                                      iconRightName="REARRANGE.svg"
                                      text={mission.name}
                                      leftChild={
                                        <React.Fragment>
                                          <InputField
                                            type="checkbox"
                                            className="mission-index--sidebar--item-check"
                                            checked={mission.status === 'active'}
                                            eventHandler={() => this.updateMissionStatus(index)}
                                          />
                                        </React.Fragment>
                                      }
                                      selected={selectedMissionId === mission.id}
                                      onClick={(_) => this.handleListClick(mission.id)}
                                    />
                                  </div>
                                )}
                              </Draggable>
                            ))}
                            {provided.placeholder}
                          </div>
                        )}
                      </Droppable>
                    </DragDropContext>
                  </React.Fragment>
                }
              </div>
            </React.Fragment>
          }
        >
          {selectedMission !== null &&
            <div className="mission-index--view">
              <div className="mission-index--view--logo">
                <img
                  src={selectedMission.imagePreview}
                />
              </div>

              <div className="mission-index--view--info">
                <div className="mission-index--view--info--item">
                  <div className="mission-index--view--info--item--name">
                    name
                  </div>
                  <div className="mission-index--view--info--item--value">
                    {selectedMission.name}
                  </div>
                </div>

                <div className="mission-index--view--info--item">
                  <div className="mission-index--view--info--item--name">
                    subtitle
                  </div>
                  <div className="mission-index--view--info--item--value">
                    {selectedMission.subtitle}
                  </div>
                </div>

                <div className="mission-index--view--info--item">
                  <div className="mission-index--view--info--item--name">
                    description
                  </div>
                  <div className="mission-index--view--info--item--value">
                    {selectedMission.description}
                  </div>
                </div>
              </div>

              <div className="mission-index--view--link">
                <span>Please select projects you want to display within selected mission:</span>
                <a
                  href={`/missions/${selectedMission.id}/edit`}
                >
                  edit mission
                </a>
              </div>
            </div>
          }

          {selectedMission !== null && selectedMission.projects.length > 0 &&
            <div className="mission-index--projects">
              {selectedMission.projects.map((p, index) =>
                <div key={p.id} className="mission-index--project-single">
                  <div className="mission-index--project-single--left">
                    <InputField
                      type="checkbox"
                      className="mission-index--project-single--check"
                      checked={p.status === 'active'}
                      eventHandler={() => this.updateProjectStatus(index)}
                    />
                    {p.title}
                  </div>
                  <Icon name="iconDropDownCopy.svg" className="styleguide-index--icon mission-index--project-dropdown" />
                </div>
              )}
            </div>
          }

        </Layout>
      </React.Fragment>
    )
  }
}

MissionIndex.propTypes = {
  missions : PropTypes.array.isRequired,
  csrfToken: PropTypes.string.isRequired
}
MissionIndex.defaultProps = {
  missions : [],
  csrfToken: '00'
}
