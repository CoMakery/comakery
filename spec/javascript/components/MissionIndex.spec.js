import React from 'react'
import { mount } from 'enzyme'
import MissionIndex from 'components/MissionIndex'

describe('MissionIndex', () => {
  it('renders correctly without props', () => {
    const wrapper = mount(<MissionIndex />)

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.exists('.mission-index')).toBe(true)
    expect(wrapper.exists('.mission-index--sidebar')).toBe(true)
    expect(wrapper.exists('.mission-index--sidebar SidebarItemBold')).toBe(true)
    expect(wrapper.find('.mission-index--sidebar SidebarItemBold').props().iconLeftName).toBe('MISSION/WHITE.svg')
    expect(wrapper.find('.mission-index--sidebar SidebarItemBold').props().iconRightName).toBe('PLUS.svg')
    expect(wrapper.exists('.mission-index--sidebar--info')).toBe(false)
    expect(wrapper.exists('.mission-index--view')).toBe(false)
  })

  it('renders correctly with missions', () => {
    const missions = [
      {
        'id'          : '0',
        'name'        : 'Mission1',
        'tokenId'     : '1',
        'subtitle'    : 'Title1',
        'description' : 'Description1',
        'status'      : 'active',
        'displayOrder': 1,
        'logoPreview' : '/logo1.png',
        'imagePreview': '/image1.png',
        'tokenName'   : 'test1',
        'tokenSymbol' : 'SYM1',
        'projects'    : []
      },
      {
        'id'          : '1',
        'name'        : 'Mission2',
        'tokenId'     : '2',
        'subtitle'    : 'Title2',
        'description' : 'Description2',
        'status'      : 'passive',
        'displayOrder': 2,
        'logoPreview' : '/logo2.png',
        'imagePreview': '/image2.png',
        'tokenName'   : 'test2',
        'tokenSymbol' : 'SYM2',
        'projects'    : [
          {
            'id'    : 1,
            'title' : 'project1',
            'status': 'active'
          },
          {
            'id'    : 2,
            'title' : 'project2',
            'status': 'passive'
          },
        ]
      },
    ]
    const wrapper = mount(<MissionIndex missions={missions} />)

    expect(wrapper.find('.mission-index--sidebar--info').text()).toBe('Please select or rearrange missions you want to display on landing page:')

    expect(wrapper.exists('SidebarItem[iconRightName="REARRANGE.svg"]')).toBe(true)
    expect(wrapper.exists('SidebarItem[text="Create a Mission"]')).toBe(true)

    expect(wrapper.exists('SidebarItem[text="Mission1"]')).toBe(true)
    expect(wrapper.find('SidebarItem[text="Mission1"]').exists('InputField[type="checkbox"][checked]')).toBe(true)

    expect(wrapper.exists('SidebarItem[text="Mission2"]')).toBe(true)
    expect(wrapper.find('SidebarItem[text="Mission2"]').exists('InputField[type="checkbox"][checked=false]')).toBe(true)
  })

  it('displays correct mission details on sidebar item click', () => {
    const missions = [
      {
        'id'          : '0',
        'name'        : 'Mission1',
        'tokenId'     : '1',
        'subtitle'    : 'Title1',
        'description' : 'Description1',
        'status'      : 'active',
        'displayOrder': 1,
        'logoPreview' : '/logo1.png',
        'imagePreview': '/image1.png',
        'tokenName'   : 'test1',
        'tokenSymbol' : 'SYM1',
        'projects'    : []
      },
      {
        'id'          : '1',
        'name'        : 'Mission2',
        'tokenId'     : '2',
        'subtitle'    : 'Title2',
        'description' : 'Description2',
        'status'      : 'passive',
        'displayOrder': 2,
        'logoPreview' : '/logo2.png',
        'imagePreview': '/image2.png',
        'tokenName'   : 'test2',
        'tokenSymbol' : 'SYM2',
        'projects'    : [
          {
            'id'    : 1,
            'title' : 'project1',
            'status': 'active'
          },
          {
            'id'    : 2,
            'title' : 'project2',
            'status': 'passive'
          },
        ]
      },
    ]
    const wrapper = mount(<MissionIndex missions={missions} />)

    expect(wrapper.exists('.mission-index--view')).toBe(false)

    wrapper.find('SidebarItem[text="Mission1"]').simulate('click')
    expect(wrapper.find('.mission-index--view').text()).toMatch(/Mission1/)
    expect(wrapper.exists('.mission-index--projects')).toBe(false)

    wrapper.find('SidebarItem[text="Mission2"]').simulate('click')
    expect(wrapper.find('.mission-index--view').text()).not.toMatch(/Mission1/)
    expect(wrapper.find('.mission-index--view').text()).toMatch(/Mission2/)
    expect(wrapper.exists('.mission-index--projects')).toBe(true)
    expect(wrapper.find('.mission-index--project-single--left').first().exists('InputField[type="checkbox"][checked]')).toBe(true)
    expect(wrapper.find('.mission-index--project-single--left').last().exists('InputField[type="checkbox"][checked=false]')).toBe(true)
  })
})
