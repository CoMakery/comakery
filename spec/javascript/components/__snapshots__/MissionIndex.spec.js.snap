// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissionIndex renders correctly without props 1`] = `
<MissionIndex
  csrfToken="00"
  missions={Array []}
>
  <Layout
    category="Missions > "
    className="mission-index"
    customTitle={null}
    hasBackButton={false}
    sidebar={
      <React.Fragment>
        <div
          className="mission-index--sidebar"
        >
          <SidebarItemBold
            className="mission-index--sidebar--item__bold"
            iconLeftName="MISSION/WHITE.svg"
            iconRightName="PLUS.svg"
            onClick={[Function]}
            text="Create a Mission"
          />
        </div>
      </React.Fragment>
    }
    subfooter={null}
    title="Configure mission’s landing page"
  >
    <div
      className="layout mission-index"
    >
      <div
        className="layout--content"
      >
        <div
          className="layout--content--title"
        >
          Configure mission’s landing page
          <span
            className="layout--content--title--category"
          >
            Missions &gt; 
          </span>
        </div>
        <hr
          className="layout--content--hr"
        />
        <div
          className="layout--content--wrapper"
        >
          <div
            className="layout--content--sidebar"
          >
            <div
              className="mission-index--sidebar"
            >
              <SidebarItemBold
                className="mission-index--sidebar--item__bold"
                iconLeftName="MISSION/WHITE.svg"
                iconRightName="PLUS.svg"
                onClick={[Function]}
                text="Create a Mission"
              >
                <SidebarItem
                  className="sidebar-item__bold mission-index--sidebar--item__bold"
                  iconLeftName="MISSION/WHITE.svg"
                  iconLeftUrl=""
                  iconRightName="PLUS.svg"
                  onClick={[Function]}
                  selected={false}
                  text="Create a Mission"
                >
                  <div
                    className="sidebar-item--wrapper"
                    onClick={[Function]}
                  >
                    <div
                      className="sidebar-item sidebar-item__bold mission-index--sidebar--item__bold"
                    >
                      <span
                        className="sidebar-item--icon-left"
                      >
                        <Icon
                          className=""
                          name="MISSION/WHITE.svg"
                        >
                          <img
                            className="icon icon__MISSION/WHITE-svg"
                            src="test-file-stub"
                          />
                        </Icon>
                      </span>
                      <span
                        className="sidebar-item--text"
                      >
                        Create a Mission
                      </span>
                      <span
                        className="sidebar-item--icon-right"
                      >
                        <Icon
                          className=""
                          name="PLUS.svg"
                        >
                          <img
                            className="icon icon__PLUS-svg"
                            src="test-file-stub"
                          />
                        </Icon>
                      </span>
                      <div
                        className="sidebar-item--left-border"
                      />
                    </div>
                  </div>
                </SidebarItem>
              </SidebarItemBold>
            </div>
          </div>
          <div
            className="layout--content--content__sidebared"
          />
        </div>
      </div>
    </div>
  </Layout>
</MissionIndex>
`;
